File: packages/sdk-typescript/src/transport.ts
/**
 * GopherHole Transport Layer
 *
 * Defines the Transport interface and implementations for HTTP, WebSocket, and Auto modes.
 * The transport handles sending JSON-RPC requests to the hub — connection lifecycle
 * and push events are managed separately by the GopherHole class.
 */

export type TransportMode = 'http' | 'ws' | 'auto';

/**
 * Transport interface for sending JSON-RPC requests to the hub.
 */
export interface Transport {
  /** Send a JSON-RPC request and return the parsed result */
  request<T>(method: string, params: Record<string, unknown>, timeoutMs?: number): Promise<T>;
  /** Whether this transport is currently able to send requests */
  readonly isOpen: boolean;
}

/**
 * HTTP Transport — sends JSON-RPC requests via HTTP POST to /a2a.
 * Always available, no connection required.
 */
export class HttpTransport implements Transport {
  constructor(
    private apiUrl: string,
    private apiKey: string,
    private defaultTimeout: number,
  ) {}

  get isOpen(): boolean {
    return true; // HTTP is always available
  }

  async request<T>(method: string, params: Record<string, unknown>, timeoutMs?: number): Promise<T> {
    const timeout = timeoutMs ?? this.defaultTimeout;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    try {
      const response = await fetch(`${this.apiUrl}/a2a`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          jsonrpc: '2.0',
          method,
          params,
          id: Date.now(),
        }),
        signal: controller.signal,
      });

      const data = await response.json();

      if (data.error) {
        throw new Error(data.error.message || 'RPC error');
      }

      return data.result as T;
    } catch (err) {
      if (err instanceof Error && err.name === 'AbortError') {
        throw new Error(`Request timeout after ${timeout}ms`);
      }
      throw err;
    } finally {
      clearTimeout(timeoutId);
    }
  }
}

/**
 * WebSocket Transport — sends JSON-RPC requests as frames over an existing WebSocket connection.
 * Requires an open WebSocket connection. Falls back to HTTP if wsFallback is enabled.
 */
export class WsTransport implements Transport {
  private pendingRequests = new Map<number, {
    resolve: (value: unknown) => void;
    reject: (reason: Error) => void;
    timer: ReturnType<typeof setTimeout>;
  }>();
  private requestId = 0;
  private httpFallback: HttpTransport | null;

  constructor(
    private getWs: () => WebSocket | null,
    private defaultTimeout: number,
    wsFallback: boolean,
    apiUrl: string,
    apiKey: string,
  ) {
    this.httpFallback = wsFallback ? new HttpTransport(apiUrl, apiKey, defaultTimeout) : null;
  }

  get isOpen(): boolean {
    const ws = this.getWs();
    return ws?.readyState === 1;
  }

  /**
   * Handle an incoming WebSocket message. Called by the GopherHole class when a
   * message arrives on the WebSocket. Returns true if the message was a JSON-RPC
   * response that was consumed, false otherwise.
   */
  handleMessage(data: Record<string, unknown>): boolean {
    if (data.jsonrpc === '2.0' && data.id != null && (data.result !== undefined || data.error !== undefined)) {
      const pending = this.pendingRequests.get(data.id as number);
      if (pending) {
        this.pendingRequests.delete(data.id as number);
        clearTimeout(pending.timer);
        if (data.error) {
          pending.reject(new Error((data.error as { message: string }).message || 'RPC error'));
        } else {
          pending.resolve(data.result);
        }
        return true;
      }
    }
    return false;
  }

  async request<T>(method: string, params: Record<string, unknown>, timeoutMs?: number): Promise<T> {
    const ws = this.getWs();
    if (!ws || ws.readyState !== 1) {
      if (this.httpFallback) {
        return this.httpFallback.request<T>(method, params, timeoutMs);
      }
      throw new Error('WebSocket not connected. Call connect() first or enable wsFallback.');
    }

    const timeout = timeoutMs ?? this.defaultTimeout;
    const id = ++this.requestId;

    return new Promise<T>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pendingRequests.delete(id);
        reject(new Error(`Request timeout after ${timeout}ms`));
      }, timeout);

      this.pendingRequests.set(id, {
        resolve: resolve as (value: unknown) => void,
        reject,
        timer,
      });

      ws.send(JSON.stringify({
        jsonrpc: '2.0',
        id,
        method,
        params,
      }));
    });
  }

  /** Clean up pending requests on disconnect */
  cleanup(): void {
    for (const [id, pending] of this.pendingRequests) {
      clearTimeout(pending.timer);
      pending.reject(new Error('WebSocket disconnected'));
    }
    this.pendingRequests.clear();
  }
}

/**
 * Auto Transport — uses HTTP for RPC requests (same as current SDK behaviour).
 * This is the default transport that preserves backwards compatibility.
 */
export class AutoTransport implements Transport {
  private http: HttpTransport;

  constructor(apiUrl: string, apiKey: string, defaultTimeout: number) {
    this.http = new HttpTransport(apiUrl, apiKey, defaultTimeout);
  }

  get isOpen(): boolean {
    return true; // HTTP is always available
  }

  async request<T>(method: string, params: Record<string, unknown>, timeoutMs?: number): Promise<T> {
    return this.http.request<T>(method, params, timeoutMs);
  }
}

File: packages/sdk-python/sdk-python/gopherhole/transport.py
"""
GopherHole Transport Layer

Defines the Transport protocol and implementations for HTTP, WebSocket, and Auto modes.
The transport handles sending JSON-RPC requests to the hub.
"""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Any, Callable, Literal, Optional, Protocol

import httpx

logger = logging.getLogger(__name__)

TransportMode = Literal["http", "ws", "auto"]


class Transport(Protocol):
    """Protocol for sending JSON-RPC requests to the hub."""

    async def request(self, method: str, params: dict[str, Any], timeout: Optional[float] = None) -> Any:
        """Send a JSON-RPC request and return the parsed result."""
        ...

    @property
    def is_open(self) -> bool:
        """Whether this transport is currently able to send requests."""
        ...

    async def close(self) -> None:
        """Clean up resources."""
        ...


class HttpTransport:
    """
    HTTP Transport — sends JSON-RPC requests via HTTP POST to /a2a.
    Always available, no connection required.
    """

    def __init__(self, api_url: str, api_key: str, default_timeout: float) -> None:
        self._api_url = api_url
        self._api_key = api_key
        self._default_timeout = default_timeout
        self._http: Optional[httpx.AsyncClient] = None
        self._request_id = 0

    @property
    def is_open(self) -> bool:
        return True

    def _ensure_http(self) -> httpx.AsyncClient:
        if self._http is None:
            self._http = httpx.AsyncClient(
                base_url=self._api_url,
                headers={
                    "Authorization": f"Bearer {self._api_key}",
                    "A2A-Version": "1.0",
                },
                timeout=self._default_timeout,
            )
        return self._http

    async def request(self, method: str, params: dict[str, Any], timeout: Optional[float] = None) -> Any:
        http = self._ensure_http()
        self._request_id += 1

        response = await http.post(
            "/a2a",
            json={
                "jsonrpc": "2.0",
                "method": method,
                "params": params,
                "id": self._request_id,
            },
            timeout=timeout or self._default_timeout,
        )
        response.raise_for_status()

        data = response.json()
        if "error" in data:
            raise Exception(data["error"].get("message", "RPC error"))

        return data["result"]

    async def close(self) -> None:
        if self._http:
            await self._http.aclose()
            self._http = None


class WsTransport:
    """
    WebSocket Transport — sends JSON-RPC requests as frames over an existing WebSocket.
    Requires an open WebSocket connection. Falls back to HTTP if ws_fallback is enabled.
    """

    def __init__(
        self,
        get_ws: Callable[[], Any],
        default_timeout: float,
        ws_fallback: bool,
        api_url: str,
        api_key: str,
    ) -> None:
        self._get_ws = get_ws
        self._default_timeout = default_timeout
        self._http_fallback: Optional[HttpTransport] = (
            HttpTransport(api_url, api_key, default_timeout) if ws_fallback else None
        )
        self._pending: dict[int, asyncio.Future] = {}
        self._request_id = 0

    @property
    def is_open(self) -> bool:
        ws = self._get_ws()
        return ws is not None and ws.open

    def handle_message(self, data: dict[str, Any]) -> bool:
        """
        Handle an incoming WebSocket message. Returns True if the message was
        a JSON-RPC response that was consumed, False otherwise.
        """
        if data.get("jsonrpc") == "2.0" and data.get("id") is not None and ("result" in data or "error" in data):
            msg_id = data["id"]
            future = self._pending.pop(msg_id, None)
            if future and not future.done():
                if "error" in data:
                    future.set_exception(Exception(data["error"].get("message", "RPC error")))
                else:
                    future.set_result(data["result"])
                return True
        return False

    async def request(self, method: str, params: dict[str, Any], timeout: Optional[float] = None) -> Any:
        ws = self._get_ws()
        if ws is None or not ws.open:
            if self._http_fallback:
                return await self._http_fallback.request(method, params, timeout)
            raise ConnectionError("WebSocket not connected. Call connect() first or enable ws_fallback.")

        self._request_id += 1
        msg_id = self._request_id
        t = timeout or self._default_timeout

        loop = asyncio.get_event_loop()
        future: asyncio.Future = loop.create_future()
        self._pending[msg_id] = future

        await ws.send(json.dumps({
            "jsonrpc": "2.0",
            "id": msg_id,
            "method": method,
            "params": params,
        }))

        try:
            return await asyncio.wait_for(future, timeout=t)
        except asyncio.TimeoutError:
            self._pending.pop(msg_id, None)
            raise TimeoutError(f"Request timeout after {t}s")

    def cleanup(self) -> None:
        """Clean up pending requests on disconnect."""
        for future in self._pending.values():
            if not future.done():
                future.set_exception(ConnectionError("WebSocket disconnected"))
        self._pending.clear()

    async def close(self) -> None:
        self.cleanup()
        if self._http_fallback:
            await self._http_fallback.close()


class AutoTransport:
    """
    Auto Transport — uses HTTP for RPC requests (same as current SDK behaviour).
    This is the default transport that preserves backwards compatibility.
    """

    def __init__(self, api_url: str, api_key: str, default_timeout: float) -> None:
        self._http = HttpTransport(api_url, api_key, default_timeout)

    @property
    def is_open(self) -> bool:
        return True

    async def request(self, method: str, params: dict[str, Any], timeout: Optional[float] = None) -> Any:
        return await self._http.request(method, params, timeout)

    async def close(self) -> None:
        await self._http.close()
